### PR TITLE
Move def.json parsing to happen after class discovery

### PR DESCRIPTION
**Pull request by tzz**
@kacf @nickanderson fixes bug in class discovery and adds to acceptance test

### CODE REVIEW

**Comment by nickanderson** [comment]
+100

**Comment by nickanderson** [comment]
this will need picked to 3.7.x and 3.8.x

**Comment by tzz** [comment]
`/me is an idiot`

**Comment by kacf** [comment]
trigger build

File: tests/acceptance/00_basics/def.json/classes.cf.json [hunk 1]
```diff
@@ -2,6 +2,8 @@
  "classes":
  {
   "test_class_29665402e2b4331f10b8d767b512cd916eeb5db9": [ "one", "of", "these", "matches_classmatch", "any" ],
+  "test_class_29665402e2b4331f10b8d767b512cd916eeb5db9_2": [ "cfengine" ],
+  "test_class_29665402e2b4331f10b8d767b512cd916eeb5db9_3": "$(sys.os)",
```
Comment (kacf) [inline tests/acceptance/00_basics/def.json/classes.cf.json]: Right, `$(sys.os)` doesn't actually match the corresponding class on some OSes. Silly, I know, but that's how it is. I think the `"cfengine"` test is sufficient though. Maybe remove this one?

Comment (kacf) [inline tests/acceptance/00_basics/def.json/classes.cf.json]: I submitted a new one here: https://github.com/cfengine/core/pull/2565

Comment (tzz) [inline tests/acceptance/00_basics/def.json/classes.cf.json]: That's fine, sorry, I didn't remember that inconsistency...

**Comment by kacf** [comment]
Superseded by #2565.

**Pull request closed without merging**